README.md
# Pairing Matrix

<img src="https://github.com/sumanmaity112/pairing-matrix/blob/main/screenshots/pairingMatrix.gif" height="500" alt="pairing-matrix">

## Changelog

Changelog can be found [here](https://github.com/sumanmaity112/pairing-matrix/blob/main/CHANGELOG.md)

## Usage

Pairing matrix can be either used as out of box docker image or can be used as library to integrate with other application.

### As Docker

Pairing Matrix docker image can be found on [dockerhub](https://hub.docker.com/r/sumanmaity112/pairing-matrix). You can run following
command to start pairing-matrix docker

```shell
docker run -p <host port>:8080 -e CONFIG_PATH="<config json path>" --name pairing-matrix pairing-matrix:<tag>
```

Once you start the docker image, it'll expose the following `http://localhost:<host port>?pull-data=<pull data>&since-days=<since days>&aggregate-by=<aggregate by>`

- `pull-data` if set to **true** then latest data will be pulled from github repos and then it'll create the matrix
- `since-days` by default it's set to **14** days but can be set to any number of days
- `aggregate-by` defines the aggregation logic needs to use while generating pairing matrix. It can be either `issue` or `days`. By default, it makes use of **issue** based aggregation

**Note** For config please check [here](https://github.com/sumanmaity112/pairing-matrix/tree/main/packages/server#config-file)

### As Library

To be able to design your own pairing matrix or include with your existing application, the core logic can be found as
npm package.

- [pairing-matrix-engine](https://www.npmjs.com/package/pairing-matrix-engine) is responsible to collect pairing data from GitHub commits
- [pairing-matrix-charts](https://www.npmjs.com/package/pairing-matrix-charts) is responsible to create D3 charts for given pairing information. It can be used with most of the popular javascript frontend library.
- [pairing-matrix-vue-visualiser](https://www.npmjs.com/package/pairing-matrix-vue-visualiser) is responsible to visualise pairing information. Internally it makes use of [pairing-matrix-charts](https://www.npmjs.com/package/pairing-matrix-charts). Currently, it can be only used with [Vuejs](https://vuejs.org/)

## Developers Guide

### Prerequisites

- [talisman](https://github.com/thoughtworks/talisman)
- [Node.js v16.13.1](https://nodejs.org)
- [nvm](https://github.com/nvm-sh/nvm)
- [ShellCheck](https://www.shellcheck.net)

### Project setup

```shell
./run.sh bootstrap
```

#### Compiles and minifies for production

```shell
yarn build
```

### Fix Lint issues

```shell
./run.sh format
```

### Commits

This repository follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/). Therefor whenever you are committing the changes make sure use proper **type**.

#### Types

- **feat** for a new feature for the user, not a new feature for build script. Such commit will trigger a release bumping a MINOR version.
- **fix** for a bug fix for the user, not a fix to a build script. Such commit will trigger a release bumping a PATCH version.
- **perf** for performance improvements. Such commit will trigger a release bumping a PATCH version.
- **docs** for changes to the documentation.
- **style** for formatting changes, missing semicolons, etc.
- **refactor** for refactoring production code, e.g. renaming a variable.
- **test** for adding missing tests, refactoring tests; no production code change.
- **build** for updating build configuration, development tools or other changes irrelevant to the user.

> **_NOTE:_** Add ! just after type to indicate breaking changes


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.0](https://github.com/sumanmaity112/pairing-matrix/compare/0.3.0...1.0.0) (2022-06-06)

### Features 

* Display pair recommendations ([1770cd](https://github.com/sumanmaity112/pairing-matrix/commit/1770cd62a0c8b33beb0f0be50c8d640a1cc148e9))
* Display data unavailable info message ([718d1e](https://github.com/sumanmaity112/pairing-matrix/commit/718d1ee1fdbb788791cec0acac756812a23507f6))

### Bug Fixes 

* Update vue-visualiser version to fix duplicate color code in chord chart ([eca99c](https://github.com/sumanmaity112/pairing-matrix/commit/eca99c4b1b014434e4f16fa18e4f651dc5e2071d))



## [0.3.0](https://github.com/sumanmaity112/pairing-matrix/compare/0.2.0...0.3.0) (2022-05-23)

### ⚠ BREAKING CHANGES

* Rename aggregate by date to aggregate by days ([d262ee](https://github.com/sumanmaity112/pairing-matrix/commit/d262eeae154bc05c5a7896fa758297f038924fee))

### Features

* Change default aggregation to days ([a7d870](https://github.com/sumanmaity112/pairing-matrix/commit/a7d8707de5c604e25c4ad5f182c4939699fe8660))


## [0.2.0](https://github.com/sumanmaity112/pairing-matrix/compare/0.1.1...0.2.0) (2022-05-08)

### ⚠ BREAKING CHANGES

* Allow user to configure pairing matrix using UI instead of url params ([18ec46](https://github.com/sumanmaity112/pairing-matrix/commit/18ec46c6ae1bba3d23fddeb44d7069aac992254d))

### Features

* Allow ability to visualise pairing matrix data in tabular form ([512bbd](https://github.com/sumanmaity112/pairing-matrix/commit/512bbd09800d087724f47eeb53908d2081d18e4c))
* Introduce TabularPairingChart ([cfc22c](https://github.com/sumanmaity112/pairing-matrix/commit/cfc22c4c186e698e521b4c09ec689e251e9bd9b6))
* Display spinner when page is loading, also display alert if something goes wrong ([6e0d75](https://github.com/sumanmaity112/pairing-matrix/commit/6e0d752280e29c601d4bed07382521e19f8389f0))
* Move config as side bar instead of top bar ([baaf13](https://github.com/sumanmaity112/pairing-matrix/commit/baaf132d2249fc1bc7267eab759a74b4ea5ab323))



## [0.1.1](https://github.com/sumanmaity112/pairing-matrix/compare/0.1.0...0.1.1) (2022-04-11)

### Bug Fixes

* Use linux image as alpine is not able to access github.com ([cbcf71](https://github.com/sumanmaity112/pairing-matrix/commit/cbcf71655d7584f54f02fa718067940121eb6d95))


## [0.1.0] (2022-04-10)

### Features

* Introduce docker image for pairing matrix application ([12b7cb](https://github.com/sumanmaity112/pairing-matrix/commit/12b7cba7cb52646bc4a2e7043df0483a133e8789))
* Introduce server to serve pairing matrix application ([beade4](https://github.com/sumanmaity112/pairing-matrix/commit/beade41e5721abeb8b81f70fb7f67129deeacf70))
* Introduce frontend application for pairing matrix ([7467aa](https://github.com/sumanmaity112/pairing-matrix/commit/7467aa586ab0372b0d29baccc24b53bc36ff0638))


server/app.js
import express from "express";
import * as path from "path";
import { fileURLToPath } from "url";
import { readFileSync } from "fs";
import _ from "lodash";
import PairingMatrixGenerator from "pairing-matrix-engine";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const config = JSON.parse(readFileSync(process.env.CONFIG_PATH, "utf8"));

const pairingMatrixGenerator = new PairingMatrixGenerator(
  config.username,
  config.repos,
  config.basePath,
  config.sshIdentityFilePath
);

const app = express();

app.use(express.urlencoded());
app.use(express.json());

app.use(express.static(path.join(__dirname, "public")));

app.get("/", (req, res) =>
  res.sendFile(path.join(__dirname, "public/index.html"))
);

const getAggregation = (aggregateBy) => {
  if (
    _.isEmpty(aggregateBy) ||
    aggregateBy === PairingMatrixGenerator.AGGREGATE_BY_ISSUE
  )
    return PairingMatrixGenerator.AGGREGATE_BY_ISSUE;
  else if (aggregateBy === PairingMatrixGenerator.AGGREGATE_BY_DATE)
    return PairingMatrixGenerator.AGGREGATE_BY_DATE;
  return null;
};

app.get(
  "/api/pair-matrix",
  (
    {
      query: {
        "since-days": sinceDays,
        "pull-data": pullData,
        "aggregate-by": aggregateBy,
      },
    },
    res,
    next
  ) => {
    const aggregation = getAggregation(aggregateBy);
    if (!aggregation) return res.sendStatus(400);
    return pairingMatrixGenerator
      .generatePairingMatrix(
        _.isEmpty(sinceDays) ? 14 : sinceDays,
        pullData === "true",
        aggregation
      )
      .then((matrix) => res.json(matrix))
      .catch(next);
  }
);

app.get("/api/sync", (req, res, next) => {
  return pairingMatrixGenerator
    .fetchRepos()
    .then(() => res.sendStatus(200))
    .catch(next);
});

app.use((err, req, res, next) => {
  console.error(err);
  res.sendStatus(500);
});

export default app;
